# DTSC 580: Data Manipulation
## Assignment: LEGO Project
### Name: Nash Lawrence

![LegoSchema](Lego_Schema.PNG)

## Overview

In this assignment, you will load the various tables from the [Rebrickable](https://rebrickable.com/) website and perform analysis on the data.  Rebrickable is a fantastic website that is used to show you which LEGO sets you can build from the sets and parts you already own.  They will even show you which parts you are missing to complete the new set and suggest sets that you could purchase to get those parts.  

## Data

The data has been downloaded directly from their website and includes files for the tables shown in the schema above.  Your first task will be to go through the data and review the schema to understand how the tables all fit together.  There is no data dictionary that I am aware of, but you should be able to understand this data by reviewing the files and schema and visiting their [Download](https://rebrickable.com/downloads/) page and [API](https://rebrickable.com/api/) documentation.  **Note that you must use the files downloaded from Brightspace for this assignment and not download new files from the Rebrickable website.**

## Assignment

In this assignment, you will create a number of functions to complete certain tasks.  This is not needed if you were working on this project on your own, but since we want to see if you can select the data via Pandas/Numpy, we will use those functions in the automatic grading in CodeGrade.  Without this, students could try to look up the answers in the `csv` files instead of wrangling the data with Python/Pandas, which would defeat the purpose of this class.

In addition, we would strongly suggest that you merge data sets as appropriate for the given task.  For example, you might merge the `sets` and `themes` DataFrames to answer one question, and then merge `sets`, `inventories`, `inventory_minifigs`, and `minifigs` to answer another.  This will make your code much easier to write and will be key in working with data across multiple tables.  In fact, I was able to create each function with only one line of code because of the merging of certain data sets.  Also note that you may not use all the tables in this assignment.  This is also an important skill for a Data Scientist to know when to use data and what data is unnecessary.

When complete, save your notebook as `lego.ipynb` and submit to CodeGrade for automatic grading.

## Note

<u>Show Work</u>

Remember that you must show your work.  Students submissions are spot checked manually throughout the term to verify that they are not hard coding the answer from looking only in the file or in CodeGrade's expected output.  If this is seen, the student's answer will be manually marked wrong and their grade will be changed to reflect this. 

For example, if the answer to Q1, the mean of a specific column, is 22:
```
# correct way
Q1 = df['column_name'].mean()

# incorrect way
Q1 = 22 
```

Let's start out by importing some standard imports.

In [1]:
# common imports
import pandas as pd
import numpy as np

# Do not change this option; This allows the CodeGrade auto grading to function correctly
pd.set_option('display.max_columns', None)

## Import Data

Load the various Lego datasets below calling them the same as the name of the table as seen in the schema.  Note that you may or may not have to use all the data sets.  One of the things that all Data Scientists and Analysts have to decide when working on a project is what data is useful for a task and what data is not.  Remember to look at the schema above to see how all the data sets are connected.

The autograder in CodeGrade will have the `csv` files in the same location where the notebook is run.  Because of this, please make sure your local files are in the same location as your notebook.

In [2]:
# import Lego data sets here
# make sure that files are in the same folder as your notebook
inventories = pd.read_csv('inventories.csv')
inventory_sets = pd.read_csv('inventory_sets.csv')
sets = pd.read_csv('sets.csv')
themes = pd.read_csv('themes.csv')
inventory_minifigs = pd.read_csv('inventory_minifigs.csv')
minifigs = pd.read_csv('minifigs.csv')
inventory_parts = pd.read_csv('inventory_parts.csv')
part_categories = pd.read_csv('part_categories.csv')
parts = pd.read_csv('parts.csv')
colors = pd.read_csv('colors.csv')
part_relationships = pd.read_csv('part_relationships.csv')
elements = pd.read_csv('elements.csv')

## Exploration

Take time to look at all of the data sets using `.head()`, `.info()`, `.describe()`, etc to familiarize yourself with the data sets.  As you look at them, notice how the various tables connect together.

In [3]:
len(minifigs)

11827

## Assignment

Answer all the questions below using the available data.  There are multiple ways to get some of the answers.  I would suggest that you attempt to merge some data sets as needed to assist with getting answers.  This will make your tasks a lot easier and your code will be cleaner.

Also, you will be creating functions to get the answers to some of these questions.  This will make it much easier for you to get answers to similar questions in the future and CodeGrade will be checking your functions using different inputs.  

**Q1:** Create a function called `data_count` that takes as input the data set (for example: `sets`) and returns the total number of rows in the data (as an integer).  Nothing else should be returned.

In [4]:
def data_count(df):
    return len(df)

**Code Check:**  If you call your `data_count` function using the `sets` DataFrame, you should see that there are 18,576 sets represented in the data.

In [5]:
data_count(sets)

18576

**Q2:** What is the average number of parts for a Lego set (rounded to the nearest integer)?  Save this as `Q2`.

In [6]:
Q2 = round(sets['num_parts'].mean(), 0).astype(int)
Q2

161

**Q3:** What is the median number of parts per set (output as an integer)?  Save this as `Q3`.

In [7]:
Q3 = int(sets['num_parts'].median())
Q3

37

You should see that the median number of parts is much lower than the average number of parts per set.  This tells us that there are some Lego sets with a very high number of parts that brings the average up.

**Q4:** What is the largest number of parts that a set has? Save this as `Q4`.

In [8]:
Q4 = sets['num_parts'].max()
Q4

11695

**Q5:** Given the set with the largest number of parts calculated above, select its row from the `sets` DataFrame.  Save this as `Q5`.  Do not reset the index.

In [9]:
Q5 = sets[sets['num_parts'] == sets['num_parts'].max()]
Q5

set_num       name  year  theme_id  num_parts
3867  31203-1  World Map  2021       709      11695

**Q6:** Create a function called `select_set_row` where you pass it the `set_num` (as a string) and the function returns the respective row of the DataFrame from `sets`. Do not reset the index.

In [10]:
def select_set_row(setNum):
    return sets[sets['set_num'] == setNum]

select_set_row('31203-1')

set_num       name  year  theme_id  num_parts
3867  31203-1  World Map  2021       709      11695

**Q7:** Create a function called `select_set_numparts` that takes as input a specific number of parts (as an integer) and selects all the rows in the `sets` DataFrame that contain those number of parts.  Return the DataFrame sorted by `year` and then by `set_num`.  Do not reset the index.

In [11]:
def select_set_numparts(numParts):
    return sets[sets['num_parts'] == numParts].sort_values(by = ['year', 'set_num'])

**Code Check:** Call the `select_set_numparts()` function to select all sets with `0` number of parts.  There should be 3,146 rows returned.

In [12]:
select_set_numparts(43)

set_num                     name  year  theme_id  num_parts
0            001-1                    Gears  1965         1         43
3966         322-4     Educational Gear Set  1969       523         43
9866         621-2       Shell Tanker Truck  1970       418         43
16693        935-1  Roof Bricks, 33 Degrees  1973       473         43
10228        646-1       Auto Service Truck  1979        67         43
...            ...                      ...   ...       ...        ...
5714       41437-1   Mia's Jungle Play Cube  2020       494         43
928        11948-1        Carnivorous Plant  2021        22         43
3593       30539-1        Lloyd's Quad Bike  2021       435         43
9028      562104-1           Ice Cream Shop  2021       494         43
17602  BLUEGHOST-1               Blue Ghost  2021       230         43

[78 rows x 5 columns]

**Q8:** 
- The oldest set listed was made in what year?  Save the year as `Q8A` (as an integer).  
- The newest set listed was made in what year? Save the year as `Q8B` (as an integer).

In [13]:
Q8A = sets['year'].min()
Q8A

1949

In [14]:
Q8B = sets['year'].max()
Q8B

2022

**Q9:** 
- Create a function called `select_set_year` that takes as input a year (as an integer) and returns the rows from the `sets` DataFrame that matches that year.  Sort the DataFrame by `set_num`.  Do not reset the index.

In [15]:
def select_set_year(yearNum):
    return sets[sets['year'] == yearNum].sort_values(by = 'set_num')

**Q10:** 
- Create a function called `theme_by_year` that takes as input a year (as an integer) and shows the theme ids and theme names (listed in order by theme id) that were in sets that year.  
- The column names must be `id` and `name_themes` (to differentiate between the name of a theme and the name of a set) in that order.  
- The index should be reset and go from `0` to `n-1`.  
- Each theme should only be listed once even if it appeared in more than one set from that year -- duplicate themes should be based on theme id and not name since there are some themes with the same name but with a different id.
- Hint:  It will help if you were to think about merging appropriate DataFrames to help you get this answer.

In [16]:
def theme_by_year(numYear):
    join_setTheme = pd.merge(sets, themes, left_on = 'theme_id', right_on = 'id')
    #dupDrop = join_setTheme.drop_duplicates(subset = 'theme_id')
    #paramYear = dupDrop[dupDrop['year'] == numYear]
    twoVars = join_setTheme[['id', 'name_y', 'year']].sort_values(by = 'id')
    themeName = twoVars.rename(columns ={'id': 'id', 'name_y': 'name_themes'})
    themeName = themeName[themeName['year'] == numYear]
    themeName = themeName.drop_duplicates(subset = 'id')
    
    return themeName[['id', 'name_themes']].reset_index(drop = True)

theme_by_year(1960)

id   name_themes
0  371  Supplemental
1  497         Books
2  513       Classic

**Code Check:** Call the `theme_by_year()` function on the year `1960`.  Your output should look as follows:

| | id | name_themes|
|-|---|---|
|0|371|Supplemental|
|1|497|Books|
|2|513|Classic|

**Optional Practice:** Call the `theme_by_year()` function using your birth year to see the Lego themes that were popular when you were born.

In [17]:
theme_by_year(1978)

id               name_themes
0     4            Expert Builder
1    67              Classic Town
2   189            Classic Castle
3   227                 Christmas
4   235                      4.5V
5   363                      Boat
6   364  Building Set with People
7   404                Hobby Sets
8   405                 Homemaker
9   412                   Airport
10  453                   Technic
11  469                 Basic Set
12  473              Supplemental
13  501                      Gear
14  505                 Basic Set
15  632                      Town

**Q11:** 
- Create a function called `theme_by_name` that takes as input a theme name and outputs all the sets (in `set_num` order) that had that theme.  
- The DataFrame should include the following columns in this order and with these exact names: `set_num`,`name_sets`,`year`,`num_parts`, `theme_id`, and `name_themes`.  
- The index should go from `0` to `n-1`.

In [18]:
def theme_by_name(themeName):
    setTheme = pd.merge(sets, themes, left_on = 'theme_id', right_on = 'id')
    colVals = setTheme[['set_num', 'name_x', 'year', 'num_parts', 'theme_id', 'name_y']].sort_values(by = 'set_num')
    nameVals = colVals.rename(columns = {'name_x': 'name_sets', 'name_y': 'name_themes'})
    themeVals = nameVals[nameVals['name_themes'] == themeName]
    
    return themeVals.reset_index(drop = True)

**Code Check:** Call the `theme_by_name()` function using the theme name `Gear`.  You should return 1,904 sets that match that theme name.

In [19]:
theme_by_name("Gear")

set_num                                          name_sets  \
0              11138-1                            Bionicle Barraki Wallet   
1       119645111436-1                Classic Space Astronaut Plush (Red)   
2       119645111504-1               Classic Space Astronaut Plush (Blue)   
3       119645111740-1              Classic Space Astronaut Plush (White)   
4       119645111818-1               Classic Space Astronaut Plush (Pink)   
...                ...                                                ...   
1899           WATCH-3                   Watch System Voyager (UFO Space)   
1900         WIIGAME-1                              Bionicle Heroes - Wii   
1901         WIIGAME-2  Indiana Jones: The Original Adventures Video G...   
1902  WOODENELEPHANT-1  Original Wooden Elephant Pull-Along Toy and Print   
1903          ZX8000-1                               ZX 8000 LEGO Sneaker   

      year  num_parts  theme_id name_themes  
0     2008          0       501        Gear  
1     2021          0       501        Gear  
2     2021          0       501        Gear  
3     2021          0       501        Gear  
4     2021          0       501        Gear  
...    ...        ...       ...         ...  
1899  1997         29       501        Gear  
1900  2007          0       501        Gear  
1901  2008          0       501        Gear  
1902  2021          0       501        Gear  
1903  2020          0       501        Gear  

[1904 rows x 6 columns]

**Q12:** Create a function called `theme_by_setnum` that takes as input the set_num (as a string) and returns the respective theme name as a string.

In [35]:
def theme_by_setnum(setString):
    joinData = pd.merge(sets, themes, left_on = 'theme_id', right_on = 'id')
    newData = joinData[['set_num', 'name_y']]
    dataVals = newData.rename(columns = {'name_y': 'theme_name'})
    dataVals = dataVals.reset_index(drop = True)
    dataVals = dataVals[dataVals['set_num'] == setString]
    dataVals = dataVals.iloc[0]['theme_name']
    
    return dataVals

theme_by_setnum('10714-1')

'Classic'

**Code Check:** Visit the [Sets page](https://rebrickable.com/sets/) from the Rebrickable website and find a set that you like.  Call your `theme_by_setnum()` function on that set and make sure the correct theme name is output.

In [38]:
print(theme_by_setnum('75810-1'))


Stranger Things


**Millennium Falcon Prep Exercise:** Find all the sets named 'Millennium Falcon'. Of these sets, find the set with the largest number of parts.  You will use this information for later questions.

In [24]:
millFalcon = pd.merge(sets, themes, left_on = 'theme_id', right_on = 'id')
milFalc = millFalcon[millFalcon['name_x'] == 'Millennium Falcon']
milFalc.nlargest(1, 'num_parts')

set_num             name_x  year  theme_id  num_parts   id  \
3199  75192-1  Millennium Falcon  2017       171       7541  171   

                         name_y  parent_id  
3199  Ultimate Collector Series      158.0

**Q13:** 
- How many minifigs are contained in the set from the `Millennium Falcon Prep Exercise`?  Save this as `Q13` (as an integer).  
- Note that there are sometimes multiple versions of a set as shown in the `inventories` table.  This will not matter for this question as there is only one version for the respective set but this should be noted for future questions.
- Remember to show your work or your answer will manually get marked as incorrect.

In [25]:
milFalc = milFalc.rename(columns = {'id': 'idT'})
x = pd.merge(milFalc, inventories, left_on = 'set_num', right_on = 'set_num')
y = pd.merge(x, inventory_minifigs, left_on = 'id', right_on = 'inventory_id')
z = pd.merge(y, minifigs, left_on = 'fig_num', right_on = 'fig_num')
z = z[z['set_num'] == '75192-1']
Q13 = z.shape[0]
Q13

8

**Q14:** 
- Create a function called `get_minifigs` that takes as input a set number (as a string) and a version number (as an integer) and outputs all minifigs from that set / version combination.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `set_num`, `fig_num`, `fig_name`, `fig_num_parts`.  
- The index should go from 0 to n-1.

In [26]:
def get_minifigs(set_string, versNum):
    inv_invMini = pd.merge(inventories, inventory_minifigs, left_on = 'id', right_on = 'inventory_id')
    invMiniFig = pd.merge(inv_invMini, minifigs, left_on = 'fig_num', right_on = 'fig_num')
    invMiniFig = invMiniFig.rename(columns = {'name': 'fig_name', 'num_parts': 'fig_num_parts'})
    figSet = invMiniFig[(invMiniFig['set_num'] == set_string) & (invMiniFig['version'] == versNum)]
    figSet = figSet[['set_num', 'fig_num', 'fig_name', 'fig_num_parts']].sort_values(by = 'fig_num')
    
    return figSet.reset_index(drop = True)

get_minifigs('3931-1', 1)

set_num     fig_num                            fig_name  fig_num_parts
0  3931-1  fig-001549  Emma - Lavender Top, Magenta Skirt              4

**Q15:** Using the above function, output information about the minifigs from the set discussed in the `Millennium Falcon Prep Exercise`.  Save this output as `Q15`.

In [27]:
Q15 = get_minifigs('75192-1', 1)
Q15

set_num     fig_num                                           fig_name  \
0  75192-1  fig-001714                              Chewbacca, Dark Brown   
1  75192-1  fig-001810                             Finn, Black Undershirt   
2  75192-1  fig-002057                                 Rey, Dark Tan Robe   
3  75192-1  fig-002514    C-3PO, Pearl Gold, Colorful Wires, Printed Legs   
4  75192-1  fig-002544                                               BB-8   
5  75192-1  fig-002549  Princess Leia, Hoth Outfit, White Jacket, Prin...   
6  75192-1  fig-002550  Han Solo, Dark Blue Jacket, Dark Brown Legs, P...   
7  75192-1  fig-002551                               Han Solo, Old, Angry   

   fig_num_parts  
0              3  
1              4  
2              4  
3              3  
4              2  
5              4  
6              4  
7              4

**Code Check:** Make sure that your `get_minifigs()` function works with both version 1 and version 2 of set 10217-1, Harry Potter's Diagon Alley.  Both versions should have 11 minifigs in the DataFrame.

In [28]:
print(get_minifigs('10217-1', 1).shape[0])
print(get_minifigs('10217-1', 1).shape[0])

11
11


**Q16:** 
- Create a function called `minifigs_from_themes` that takes as input a theme id (as an integer) and outputs all the minifigs contained in sets with that theme.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `fig_num`, `fig_name`. 
- The index should go from 0 to n-1.  
- Exact duplicates should be removed so that only unique figs are included.

In [29]:
def minifigs_from_themes(themeID):
    set2Inv = pd.merge(sets, inventories, left_on = 'set_num', right_on = 'set_num')
    setInv_fig = pd.merge(set2Inv, inventory_minifigs, left_on = 'id', right_on = 'inventory_id')
    setMinis = pd.merge(setInv_fig, minifigs, left_on = 'fig_num', right_on = 'fig_num')
    setMinis = setMinis.rename(columns = {'name_y': 'fig_name'})
    setID = setMinis[setMinis['theme_id'] == themeID]
    themeFigs = setID[['fig_num', 'fig_name']].sort_values(by = 'fig_num')
    themeFigs = themeFigs.drop_duplicates(subset = 'fig_num')
    
    return themeFigs.reset_index(drop = True)

**Code Check:** Call the `minifigs_from_themes()` function using the theme number of the set discussed in the `Millennium Falcon Prep Exercise`.  There should be 149 minifigs output for that respective theme.

In [30]:
minifigs_from_themes(171).shape[0]

149

**Q17:** 
- Create a function called `sets_from_minifig` that takes as input a minifig number (as a string) and returns a list of sets a minifig has appeared in.  
- Return a DataFrame (sorted by `set_num` and then by `version`) with the following columns: `set_num`, `version`,`name_sets`, `fig_num`, `fig_name`.  
- The index should go from 0 to n-1.

In [31]:
def sets_from_minifig(figNum):
    sets2Inv = pd.merge(sets, inventories, left_on = 'set_num', right_on = 'set_num')
    setsInv_fig = pd.merge(sets2Inv, inventory_minifigs, left_on = 'id', right_on = 'inventory_id')
    setsMini = pd.merge(setsInv_fig, minifigs, left_on = 'fig_num', right_on = 'fig_num')
    setsMini = setsMini.rename(columns = {'name_x': 'name_sets', 'name_y': 'fig_name'})
    figEntry = setsMini[setsMini['fig_num'] == figNum]
    figSets = figEntry[['set_num', 'version', 'name_sets', 'fig_num', 'fig_name']].sort_values(by = ['set_num', 'version'])
    
    return figSets.reset_index(drop = True)

**Code Check:** Call your `sets_from_minifig()` function on one of the Lego Friends minifigs, Andrea, who has figure number: `fig-001039`.  Your function should return a DataFrame that contains the following sets:

- Andrea on the Beach
- Birthday Party
- City Park Cafe

In [32]:
sets_from_minifig('fig-001039')

set_num  version            name_sets     fig_num  \
0  30100-1        1  Andrea on the Beach  fig-001039   
1  30107-1        1       Birthday Party  fig-001039   
2   3061-1        1       City Park Café  fig-001039   

                                            fig_name  
0  Andrea - Bright Light Orange Top, Light Aqua S...  
1  Andrea - Bright Light Orange Top, Light Aqua S...  
2  Andrea - Bright Light Orange Top, Light Aqua S...

**Q18:** 

*Note:  Before you begin working on this question, please note that CodeGrade has a 1GB RAM limit.  If you see an `Exit -9` code when submitting your assignment, it is probably due to this RAM limit and previous students had an issue with this question causing too much memory usage.  The biggest thing to think about is to only use the tables that you need to answer for this question.  For instance, you do not need any of the minifigs data since this is not needed for the output and the merging of this data could cause the number of rows to "explode" in terms of the total count.  I understand that this can be annoying that we have to work around CodeGrade's limitations, but it is good practice in memory management.*

*You should only need to merge the following datasets for this answer so as to not go above the RAM limit: `themes`, `sets`, `inventories`, `inventory_parts`, `parts`, and `colors`*

- Create a function called `set_parts` that takes as input the set number (as a string) and version number (as an integer) and outputs a DataFrame with each part in the set.  
- The DataFrame should have the following columns/column names: `set_num`, `version`, `name_sets`, `year`, `name_themes`, `part_num`, `name_part`, `name_color`,`quantity`,`is_spare`.  
- The Data should be sorted by `part_num` and then `name_color`.  
- The index should go from 0 to n-1.

In [33]:
def set_parts(setNumStr, verzNum):
    mOne = pd.merge(sets, themes, left_on = 'theme_id', right_on = 'id')
    mOne = mOne.rename(columns = {'id': 'idT', 'name_x': 'name_sets', 'name_y': 'name_themes'})
    mTwo = pd.merge(mOne, inventories, left_on = 'set_num', right_on = 'set_num')
    mThree = pd.merge(mTwo, inventory_parts, left_on = 'id', right_on = 'inventory_id')
    mFour = pd.merge(mThree, parts, left_on = 'part_num', right_on = 'part_num')
    mFive = pd.merge(mFour, colors, left_on = 'color_id', right_on = 'id')
    mFive = mFive.rename(columns = {'name_x': 'name_part', 'name_y': 'name_color'})
    inpSetNum = mFive[(mFive['set_num'] == setNumStr) & (mFive['version'] == verzNum)]
    partSets = inpSetNum[['set_num', 'version', 'name_sets', 'year', 'name_themes', 'part_num', 'name_part', 'name_color',
                         'quantity', 'is_spare']].sort_values(by = ['part_num', 'name_color'])
    
    return partSets.reset_index(drop = True) 

**Code Check**: Using the `set_parts()` function, output a list of parts for the set discussed in the `Millennium Falcon Prep Exercise`. There should be a total of 730 rows in the DataFrame, which includes 676 parts and 54 spare parts. You can check the respective page on the rebrickable website to verify these results.

In [34]:
set_parts('21041-1', 2)

set_num  version            name_sets  year   name_themes part_num  \
0   21041-1        2  Great Wall of China  2018  Architecture    11211   
1   21041-1        2  Great Wall of China  2018  Architecture    14417   
2   21041-1        2  Great Wall of China  2018  Architecture    14704   
3   21041-1        2  Great Wall of China  2018  Architecture    15573   
4   21041-1        2  Great Wall of China  2018  Architecture    15573   
..      ...      ...                  ...   ...           ...      ...   
81  21041-1        2  Great Wall of China  2018  Architecture    92280   
82  21041-1        2  Great Wall of China  2018  Architecture    96874   
83  21041-1        2  Great Wall of China  2018  Architecture    98283   
84  21041-1        2  Great Wall of China  2018  Architecture    98560   
85  21041-1        2  Great Wall of China  2018  Architecture    99206   

                                            name_part         name_color  \
0          Brick Special 1 x 2 with 2 Studs on 1 Side                Tan   
1   Plate Special 1 x 2 with 5.9mm Centre Side Tow...   Dark Bluish Gray   
2           Plate Special 1 x 2 5.9mm Centre Side Cup  Light Bluish Gray   
3   Plate Special 1 x 2 with 1 Stud with Groove an...         Dark Green   
4   Plate Special 1 x 2 with 1 Stud with Groove an...                Tan   
..                                                ...                ...   
81                     Plate Special 1 x 2 [Top Clip]             Yellow   
82                      Brick and Axle Separator v2.0             Orange   
83     Brick Special 1 x 2 with Masonry Brick Profile                Tan   
84                  Slope 75° 2 x 2 x 3 [Solid Studs]         Dark Green   
85  Plate Special 2 x 2 x 0.667 with Two Studs On ...                Tan   

    quantity is_spare  
0          1        f  
1          1        f  
2          1        f  
3          2        f  
4         14        f  
..       ...      ...  
81         2        f  
82         1        f  
83        21        f  
84        13        f  
85         4        f  

[86 rows x 10 columns]

## Optional Exercises
These exercises are optional and are not graded by CodeGrade.  They can be used as extra optional practice.

- What is the median year for the sets data?  Given the oldest and newest year in the data, what does this median value tell you about whether the number of sets have increased or decreased over the years?
- What minifig has the most number of parts?
- What minifig shows up in the most number of sets?
- When was the first minifig included in a set?
- What set has the most number of minifigs
- What part is used the most for all sets? 
- What color is used the most for all parts?
- Get a list of all unique color names, sort by color name, and output as a list. 
- How many total transparent colors are in the colors DataFrame?
- Create a function called get_part_colors that takes as input a part number (as a string) and returns a DataFrame of all colors a part has appeared in.
- What part / color combination is used the most for all sets?
- What part / color combination is used the least for all sets?
- Explore some of your favorite themes and which colors are used the most for those themes.
- Plot a line graph of the average number of parts per set per year.
